splash-reward-distributor/src/engine/verifier.rs

```diff
@@ -149,56 +165,158 @@ where
             .rollback(HashSet::new(), tx_hash);
     }
+
+    fn confirm_block_slot(&mut self, block_slot: u64) {
+        self.block_slot_buffer.add(block_slot);
+    }
+
+    fn rollback_block_slot(&mut self, block_slot: u64) {
+        let rolled_back_slot = self
+            .block_slot_buffer
+            .pop_back()
+            .expect("Block slot buffer is empty");
+        assert!(
+            rolled_back_slot == block_slot,
+            "Rolled back block slot {} != confirmed block slot {}",
+            rolled_back_slot,
+            block_slot
+        );
+    }
 }
 
 #[derive(Clone, PartialEq, Eq, Debug, Serialize, Deserialize)]
 pub struct AuthorizedExecutors(pub Vec<PublicKey>);
 
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub enum TxCosignRequest {
+    Harvest(PartiallySignedCardanoTx),
+    GaugeBuffering(PartiallySignedCardanoTx),
+}
+
 #[async_trait::async_trait]
-impl<Index, PositionIndex, UHarvestIndex, Prov, Ctx> LocalVerifier<PartiallySignedCardanoTx, Transaction, Ctx>
-    for Verifier<PartiallySignedCardanoTx, Index, PositionIndex, UHarvestIndex, Prov>
+impl<Index, PositionIndex, UHarvestIndex, Prov, Ctx> LocalVerifier<TxCosignRequest, Transaction, Ctx>
+    for Verifier<TxCosignRequest, Index, PositionIndex, UHarvestIndex, Prov>
 where
-    Index: Send + Sync,
+    Index: AuthManagerIndex<FarmId, OutputRef, FinalizedTxOut>
+        + HarvestOrderIndex<OutputRef, FinalizedTxOut>
+        + Send
+        + Sync,
     PositionIndex: Accounts<OutputRef> + Send + Sync,
     UHarvestIndex: UnconfirmedHarvestTxIndex + Send + Sync,
     Prov: TxProver<SignedTxBuilder, Transaction> + Send + Sync,
     Ctx: Has<MinLovelacePerHarvest>
         + Has<DeployedScriptInfo<{ ProtocolValidator::HarvestOrder as u8 }>>
+        + Has<DeployedScriptInfo<{ ProtocolValidator::SmartFarm as u8 }>>
         + Has<NetworkId>
+        + Has<GenesisEpochStartTime>
+        + Has<PermManagerAuthPolicy>
         + Has<SplashPolicy>
         + Has<AuthorizedExecutors>
         + Has<BufferWalletScript>
         + Sync,
 {
-    async fn try_approve(&mut self, tx: &PartiallySignedCardanoTx, ctx: &Ctx) -> Option<Transaction> {
-        if let Some(proposed_harvest_tx) = <ProposedHarvestTx<OutputRef>>::try_from_ledger(tx, ctx) {
-            let withdrawals = proposed_harvest_tx.withdrawals;
-            for withdrawal in &withdrawals {
-                let order = withdrawal.order.clone();
-                if let Some(AccountReward {
-                    accumulated_amount: amount,
-                    latest_epoch_inclusive,
-                }) = self
-                    .position_index
-                    .query_account_reward(&Credential::new_pub_key(order.account_key), Epoch::from(0)) // todo: use correct epoch
-                    .await
-                {
-                    if amount != withdrawal.amount {
+    async fn try_approve(&mut self, tx: &TxCosignRequest, ctx: &Ctx) -> Option<Transaction> {
```

Comment:
Verification is now complete for both harvest-order and gauge-buffering TXs, except for `buffer_wallet` verification, which I'll finish next in DEX-962